### PR TITLE
Add docs and type annotations to `unit_buildmenu_config.lua`

### PR DESCRIPTION
**Pull request by rhys-vdw**
### Work done

> [!IMPORTANT]
> Includes changes from #3935, please review and merge that PR first.

Adds lua language server docs and type annotations to `unit_buildmenu_config.lua`.

This PR replaces #3905 which was rejected for including code cleanup and documentation changes in one PR.

#### Test steps
No functional changes

### CODE REVIEW

File: luaui/configs/buildmenu_sorting.lua [hunk 1]
```diff
@@ -769,14 +770,14 @@ local unitOrderTable = {
...
 
+---@type table<string, number>
 local newUnitOrder = {}
 for id, value in pairs(unitOrderTable) do
 	if UnitDefNames[id] then
 		newUnitOrder[UnitDefNames[id].id] = value
 	end
 end
 unitOrderTable = newUnitOrder
-newUnitOrder = nil
```
Comment (saurtron) [inline luaui/configs/buildmenu_sorting.lua:779]: thought you wanted to leave this for #3935

**Comment by saurtron** [comment]
couple things slipped from #3935 but rest looks fine!

File: luaui/configs/unit_buildmenu_config.lua [hunk 2]
```diff
@@ -3,22 +3,18 @@
...
-local isBuilder = {}
-local isFactory = {}
-local unitIconType = {}
-local isMex = {}
-local isWind = {}
-local isWaterUnit = {}
-local isGeothermal = {}
-local unitMaxWeaponRange = {}
-
-local showWaterUnits = false
```
[the remark this answers is not in the document]
Comment (rhys-vdw) [reply luaui/configs/unit_buildmenu_config.lua:21]: Other PR needs to be merged first.

**Comment by rhys-vdw** [comment]
> couple things slipped from #3935 but rest looks fine!

They didn't slip in, the changes are tightly coupled so the other PR will need to be merged first. This is branched off that.

**Comment by rhys-vdw** [comment]
> it's not very good practice to base one PR on top of another

Agreed!

I was told PRs were reviewed merged on Thursday, and instead of being merged I just got this comment:
https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3905#discussion_r1840997739

**Comment by WatchTheFort** [comment]
@rhys-vdw Can you resolve merge conflicts, then this should be good to merge

**Comment by WatchTheFort** [comment]
@rhys-vdw Why did you close this PR?

**Comment by saurtron** [comment]
> @rhys-vdw Why did you close this PR?

there's a new one https://github.com/beyond-all-reason/Beyond-All-Reason/pull/3987

**Comment by WatchTheFort** [comment]
Yes, but why?

**Pull request closed without merging**